### PR TITLE
NodeMaterial: getHash and customProgramCacheKey

### PR DESCRIPTION
**Pull request by sunag**
`Node.getHash()` or `NodeMaterial.getHash()` generate a JSON.

This approach building the string is more fast that` array.join` and `JSON.stringify`.

### CODE REVIEW

**Review by Mugen87 — APPROVED** [review]

**Comment by Mugen87** [comment]
This is great! 👍 

@mrdoob When this is merged, it should be okay to merge #19673 (which ensures `onBeforeCompile()` is called at the correct time and improves the performance of `getParameters()`).

File: examples/jsm/nodes/core/Node.js [hunk 1]
```diff
@@ -100,6 +100,28 @@ Node.prototype = {
 
 	},
 
+	getHash: function() {
```
Comment (WestLangley) [inline examples/jsm/nodes/core/Node.js:103]: There are multiple formatting issues in this method.

File: examples/jsm/nodes/materials/NodeMaterial.js [hunk 2]
```diff
@@ -91,6 +65,43 @@ Object.defineProperties( NodeMaterial.prototype, {
...
+};
+
+NodeMaterial.prototype.getHash = function () {
+
+	var hash = '{';
+
+	hash += '"vertex":' + this.vertex.getHash() + ',\n';
+	hash += '"fragment":' + this.fragment.getHash() + '\n';
+
+	hash += '}'
```
Comment (WestLangley) [inline examples/jsm/nodes/materials/NodeMaterial.js:99]: semicolon

File: examples/jsm/nodes/core/Node.js (hunk 1 above)
Comment (sunag) [reply examples/jsm/nodes/core/Node.js:103]: I did not think that formatting the JSON was necessary... you can post an example of the style?

Comment (Mugen87) [reply examples/jsm/nodes/core/Node.js:103]: I can clean things up after merging. Don't be worry about this 👍 .

**Comment by sunag** [comment]
`Node.hashProperties ` is necessary in case of variation of the shader from the node config.

**Comment by mrdoob** [comment]
@sunag ready to be merged?

**Comment by sunag** [comment]
> @sunag ready to be merged?

Yes!

File: examples/jsm/nodes/core/Node.js [hunk 3]
```diff
@@ -100,6 +102,42 @@ Node.prototype = {
...
+
+		var hash = '{';
+		var prop, obj;
+
+		for(prop in this) {
```
Comment (WestLangley) [inline examples/jsm/nodes/core/Node.js:110]: @Mugen said

>I can clean things up after merging. Don't be worry about this 

Please be consistent. If users are not expected to follow the three.js code style, then please remove the instructions to do so found in https://github.com/mrdoob/three.js/wiki/How-to-contribute-to-three.js.

>Format whitespace consistently with the rest of code base. See the [an] article Mr.doob's-Code-Style™, and the online code beautifier. It is highly recommended to use ESLint plugin and appropriate settings in your code editor.

/ping @mrdoob

Comment (sunag) [reply examples/jsm/nodes/core/Node.js:110]: @WestLangley I see no problem in fix this... so far I thought that you is referring to the json output.

**Comment by mrdoob** [comment]
Thanks!

File: examples/jsm/nodes/core/Node.js (hunk 3 above)
Comment (Mugen87) [reply examples/jsm/nodes/core/Node.js]: Sorry, just wanted to get this merged quickly :nerd_face:.

**Pull request merged**